File: ArtSourceWrapper/Tumblr.cs
using DontPanic.TumblrSharp;
using DontPanic.TumblrSharp.Client;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace ArtSourceWrapper {
    public abstract class TumblrSubmissionWrapper<T> : ISubmissionWrapper, IDeletable where T : BasePost {
		private readonly TumblrClient _client;
		public readonly T Post;
		
		public TumblrSubmissionWrapper(TumblrClient client, T post) {
			_client = client;
			Post = post ?? throw new ArgumentNullException(nameof(post));
		}

		public virtual string Title => "";
		public abstract string HTMLDescription { get; }
		public bool Mature => false;
		public bool Adult => false;
		public IEnumerable<string> Tags => Post.Tags;
		public DateTime Timestamp => Post.Timestamp;
		public string ViewURL => Post.Url;
		public virtual string ImageURL => $"https://api.tumblr.com/v2/blog/{Post.BlogName}.tumblr.com/avatar/512";
		public virtual string ThumbnailURL => $"https://api.tumblr.com/v2/blog/{Post.BlogName}.tumblr.com/avatar/128";
		public Color? BorderColor => null;
		
		public string SiteName => "Tumblr";

		public async Task DeleteAsync() {
			await _client.DeletePostAsync(Post.BlogName, Post.Id);
		}
	}

    public class TumblrPhotoPostSubmissionWrapper : TumblrSubmissionWrapper<PhotoPost> {
		public TumblrPhotoPostSubmissionWrapper(TumblrClient client, PhotoPost post) : base(client, post) { }
		
        public override string HTMLDescription => Post.Caption;
        public override string ImageURL => Post.Photo.OriginalSize.ImageUrl;
        public override string ThumbnailURL {
            get {
				foreach (var alt in Post.Photo.AlternateSizes.OrderBy(s => s.Width)) {
                    if (alt.Width < 120 && alt.Height < 120) continue;
                    return alt.ImageUrl;
                }
                return Post.Photo.OriginalSize.ImageUrl;
            }
        }
	}

	public class TumblrTextPostSubmissionWrapper : TumblrSubmissionWrapper<TextPost>, IStatusUpdate {
		public TumblrTextPostSubmissionWrapper(TumblrClient client, TextPost post) : base(client, post) { }

		public override string Title => Post.Title;
		public override string HTMLDescription => Post.Body;

		public string FullHTML => string.IsNullOrEmpty(Title)
			? HTMLDescription
			: $"<h1>{Title}</h1>{HTMLDescription}";
		public bool PotentiallySensitive => Mature || Adult;
		public bool HasPhoto => false;
		public IEnumerable<string> AdditionalLinks => Enumerable.Empty<string>();
	}
}

File: SourceWrappers/Tumblr.fs
namespace SourceWrappers

open DontPanic.TumblrSharp.Client
open DontPanic.TumblrSharp
open FSharp.Control

[<AbstractClass>]
type TumblrPostWrapper<'T when 'T :> BasePost>(client: TumblrClient, post: BasePost) =
    abstract member HTMLDescription: string with get

    interface IPostBase with
        member this.Title = ""
        member this.HTMLDescription = this.HTMLDescription
        member this.Mature = false
        member this.Adult = false
        member this.Tags = post.Tags :> seq<string>
        member this.Timestamp = post.Timestamp
        member this.ViewURL = post.Url

    interface IDeletable with
        member this.SiteName = "Tumblr"
        member this.DeleteAsync () = client.DeletePostAsync(post.BlogName, post.Id)

type TumblrPhotoPostWrapper(client: TumblrClient, post: PhotoPost) =
    inherit TumblrPostWrapper<PhotoPost>(client, post)
    override __.HTMLDescription = post.Caption

    interface IRemotePhotoPost with
        member __.ImageURL = post.Photo.OriginalSize.ImageUrl
        member __.ThumbnailURL =
            post.Photo.AlternateSizes
            |> Seq.sortBy (fun s -> s.Width)
            |> Seq.filter (fun s -> s.Width >= 120 && s.Height >= 120)
            |> Seq.map (fun s -> s.ImageUrl)
            |> Seq.append (Seq.singleton post.Photo.OriginalSize.ImageUrl)
            |> Seq.head

type TumblrTextPostWrapper(client: TumblrClient, post: TextPost) =
    inherit TumblrPostWrapper<TextPost>(client, post)
    override __.HTMLDescription = post.Body

type TumblrSourceWrapper(client: TumblrClient, blogName: string, photosOnly: bool) =
    inherit AsyncSeqWrapper()

    let blogNamesTask = lazy(
        let a = async {
            let! user = client.GetUserInfoAsync() |> Async.AwaitTask
            return user.Blogs |> Seq.map (fun b -> b.Name)
        }
        Async.StartAsTask a
    )

    override __.Name =
        if photosOnly then
            "Tumblr (images)"
        else
            "Tumblr (text + images)"
    
    override __.FetchSubmissionsInternal() = asyncSeq {
        let! blogNames = Async.AwaitTask (blogNamesTask.Force())
        let t = if photosOnly then PostType.Photo else PostType.All

        let mutable skip = 0L
        let mutable more = true
        while more do
            let! posts =
                client.GetPostsAsync(
                    blogName,
                    skip,
                    20,
                    t,
                    true) |> Async.AwaitTask

            for post in posts.Result do
                let postBlogName =
                    if not (isNull post.RebloggedRootName)
                        then post.RebloggedRootName
                        else post.BlogName
                if blogNames |> Seq.contains postBlogName then
                    match post with
                    | :? PhotoPost as photo -> yield TumblrPhotoPostWrapper(client, photo) :> IPostBase
                    | :? TextPost as text -> yield TumblrTextPostWrapper(client, text) :> IPostBase
                    | _ -> ()
            skip <- skip + posts.Result.LongLength
            more <- posts.Result.LongLength > 0L
    }

    override __.FetchUserInternal() = async {
        let blogHostname =
            if blogName.Contains(".") then
                sprintf "%s.tumblr.com" blogName
            else
                blogName

        return {
            username = blogName
            icon_url = Some (sprintf "https://api.tumblr.com/v2/blog/%s/avatar/64" blogHostname)
        }
    }